Kubelet: add garbage collection for new runtime API

feiskyer opened the pull request:
This PR adds garbage collection for new runtime API.

Note that this PR also adds `CreatedAt` and `PodSandboxId` to `ListContainers()` result.

CC @yujuhong @Random-Liu  @kubernetes/sig-node @kubernetes/sig-rktnetes

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31326)

<!-- Reviewable:end -->

feiskyer committed:
Kubelet: add garbage collection for new runtime API

yifan-gu commented:
The second LGTM

yujuhong commented:
Reviewed 9 of 9 files at r1.
Review status: 8 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 28 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPz_LIq0-uHhjzZE_F5:-KPz_LIrkQ0_wiGKb85n:bww3pqz) ([raw file](https://github.com/kubernetes/kubernetes/blob/05e0627835b3569784c24f1d4a0d9215359c3020/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L28)):*
> ```Go
> 	kubecontainer "k8s.io/kubernetes/pkg/kubelet/container"
> )
> ```

I'd suggesting writing a brief explanation on how kubelet GC policy is interpreted in this file.

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 84 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPzae45jvSsVLR3XyCM:-KPzae45jvSsVLR3XyCN:bqwzq6z) ([raw file](https://github.com/kubernetes/kubernetes/blob/05e0627835b3569784c24f1d4a0d9215359c3020/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L84)):*
> ```Go
> 
> // evictableSandboxes gets evictable pod sandboxes.
> // Evictable pod sandboxes are not active and created more than MinAge ago)
> ```

I think podsandbox should also comply to `MaxPerPodContainer` policy, so that we at least keep one exited pod (sandbox+containers) .
In the future, we'll change the GC policy to be more podsandbox friendly, but right now, I think this is the right interpretation. 

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
Review status: 6 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 28 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPz_LIq0-uHhjzZE_F5:-KQ4G0-D7_kw19OSRfl-:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/05e0627835b3569784c24f1d4a0d9215359c3020/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L28)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> I'd suggesting writing a brief explanation on how kubelet GC policy is interpreted in this file.
</details>
Done.

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 84 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPzae45jvSsVLR3XyCM:-KQ4G11AKtqxnydVwgyT:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/05e0627835b3569784c24f1d4a0d9215359c3020/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L84)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> I think podsandbox should also comply to `MaxPerPodContainer` policy, so that we at least keep one exited pod (sandbox+containers) .
> In the future, we'll change the GC policy to be more podsandbox friendly, but right now, I think this is the right interpretation. 
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
@yujuhong The latest commit rewrites the gc logic and addresses the comments. PTAL.

yujuhong commented:
> I think podsandbox should also comply to MaxPerPodContainer policy, so that we at least keep one exited pod (sandbox+containers) .
> In the future, we'll change the GC policy to be more podsandbox friendly, but right now, I think this is the right interpretation.

Hmm......I thought about it more, and I think the correct/easy way to do this is:

 1. remove containers based on `MaxContainerPerPod`
 2. only remove a sandbox if it's empty (i.e., no containers are using it).

@feiskyer WDYT? (and sorry for the churn).

feiskyer commented:
@yujuhong So this means there will be no something like `MaxSandboxPerPod` in the future. SGTM.

And the entire GC procedure will be:

* ~~removes oldest dead sandboxes for each Pod by enforcing MaxPerPodContainer~~
* remove oldest dead containers for each Pod by enforcing MaxPerPodContainer
* remove oldest dead containers by enforcing MaxContainers
* remove dead sandboxes with zero containers.

@euank @yifan-gu Do you think this is ok for rkt runtime?

euank commented:
@feiskyer because that can be expressed by CRI it'll have to be okay for any runtime that integrates that way (that is future rkt, if not current rkt).

I agree with @yujuhong about that being the logical way to do it so long as I'm reading that correctly.

I also think it might make sense to not count sandboxes as "containers" anywhere or in any of the math. Sandboxes should simply be removed immediately when there are no remaining containers for them (dead or alive) and no intention of creating new containers in them.

I think we can potentially make the code slightly simpler if we can assume that when all containers for a pod are laid out by age, each sandbox encompasses a non-overlapping contiguous segment of containers. I think that's always true, and makes it so that avoiding counting sandboxes and just auto-removing them when empty does more or less "the right thing".

It's possible the above behaviour and assumptions are implicit in your comment, but I wanted to be a little more explicit about that part.

yujuhong commented:
> I also think it might make sense to not count sandboxes as "containers" anywhere or in any of the math. Sandboxes should simply be removed immediately when there are no remaining containers in them and no intention of creating new containers in them.

Thanks for summarizing! I think this is what @feiskyer plans to do, if I read his comments correctly. 
Essentially, all the existing GC policies applies to "non-sandbox" containers, and sandboxes are only removed when all containers associated with the sandbox are removed. 
This interprets the "MaxContainers" in the GC policy a bit differently than its original intention, i.e., we don't count sandboxes at all. I think this is acceptable, and we just have to document this behavior clearly.

feiskyer commented:
> > I also think it might make sense to not count sandboxes as "containers" anywhere or in any of the math. Sandboxes should simply be removed immediately when there are no remaining containers in them and no intention of creating new containers in them.
> Thanks for summarizing! I think this is what @feiskyer plans to do, if I read his comments correctly. 
Essentially, all the existing GC policies applies to "non-sandbox" containers, and sandboxes are only removed when all containers associated with the sandbox are removed. 
This interprets the "MaxContainers" in the GC policy a bit differently than its original intention, i.e., we don't count sandboxes at all. I think this is acceptable, and we just have to document this behavior clearly.

@euank @yujuhong Exactly. Will make an update today.

feiskyer commented:
@euank @yujuhong updated. PTAL.

tmrts commented on pkg/kubelet/kuberuntime/doc.go:
```diff
@@ -0,0 +1,19 @@
+*/
+
+// Package kuberuntime contains an implementation of kubecontainer.Runtime using
+// the interface in pkg/kubelet/api/services.go.
```
no need to include `services.go` in the path, pkg name is sufficient

feiskyer replied:
Fixed.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_container.go:
```diff
@@ -232,6 +235,73 @@ func makeUID() string {
+		}
+
+		message := ""
+		if !cStatus.FinishedAt.IsZero() || cStatus.ExitCode != 0 {
```
this looks like a good place to extract a method

feiskyer replied:
This is part of #30267, but I can fix it here for avoiding rebasing other PRs.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+			continue
+		}
+
+		err := gc.client.RemovePodSandbox(sandboxID)
```
`if err := ... ; err != nil {`

feiskyer replied:
Fixed.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+// Note that it doesn't delete the container from gc.containers.
+func (gc *podGCInfo) removeContainer(containerID string) int {
+	containerIndex := -1
+	for i, c := range gc.containers {
```
`ok, err := getContainer(cid)` method extraction

feiskyer replied:
I think this method extraction is not needed since the logic is simple enough.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	if containerIndex != -1 {
+		err := gc.client.RemoveContainer(containerID)
```
no need to separate if and err assignment here

feiskyer replied:
done.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+// getEvictablePods gets evictable pods.
+// Evictable pods include sandboxes and containers which are not active and created more than MinAge ago).
```
what do you mean by it includes containers?

feiskyer replied:
I mean `contains` here.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	evictablePods := make(map[string]*podGCInfo)
+	newestGCTime := time.Now().Add(-minAge)
```
use time.Sub

feiskyer replied:
`time.Sub` returns `Duration`, while `time.Time` is expected.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+
+// GarbageCollect removes dead containers using the specified container gc policy.
+// It :
+// * gets evictable pods with belonging containers and sandboxes which are not active and created more than MinAge ago.
```
inconsistent bullet point sentence structures, use declarative for all or imperative for all and capital case for bullets

feiskyer replied:
done.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+func (m *kubeGenericRuntimeManager) GarbageCollect(gcPolicy kubecontainer.ContainerGCPolicy, allSourcesReady bool) error {
+	evictablePods, err := m.getEvictablePods(gcPolicy.MinAge)
+	if err != nil {
+		glog.Warningf("Get evictable pods failed: %v", err)
```
`getEvictablePods` to ease debugging in the future instead of `Get evictable pods`

feiskyer replied:
fixed.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,231 @@
+	glog.V(4).Infof("GarbageCollect gets pods %q for evicting", evictablePods)
+
+	// Enforce max number of dead containers/sandboxes of each pod
+	totalEvictableContainers := 0
```
extract method

feiskyer replied:
fixed.

tmrts commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -0,0 +1,139 @@
+				Namespace: "new",
+			},
+			Spec: api.PodSpec{
+				Containers: []api.Container{
```
doesn't this error out with `gofmt -s`?

feiskyer replied:
No error with gofmt. What error do you expect here?

tmrts commented:
@feiskyer LGTM after my comments

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+}
+
+// enforceMaxPerPodContainer enforces max number of dead containers of each pod and
+// returns the number of total evictable containers.
```
I don't think you're implementing the GC policy correctly. The policy says that we can keep maximum `MaxPerPodContianer` containers for each `(pod UID, container name)` tuple. It seems like you are trying to use simply `(pod UID)` as the identifier, and remove the oldest N containers in the pod regardless of what names they have.

feiskyer replied:
@yujuhong Thanks for figuring out this. Will fix.

yujuhong replied:
The reason behind this is we want to ensure there will be at least N dead container per `(pod UID, container name)` tuple to get the status/logs. That's why the "container name" is important :)

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+	glog.V(4).Infof("GarbageCollect gets pods %q for evicting", evictablePods)
+
+	// Enforce max number of dead containers of each pod
+	totalEvictableContainers := enforceMaxPerPodContainer(evictablePods, gcPolicy.MaxPerPodContainer)
```
I'm not sure why you call this variable `totalEvictableContainers`. They've been removed/evicted already, and this is just an integer...

yujuhong replied:
ah..I see you are just want to return the rest of the containers that are evictable. It's not very clear from just reading the description of the function.

feiskyer replied:
Y, should add a `Number` prefix for more clear.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+func enforceMaxPerPodContainer(evictablePods map[string]*podGCInfo, maxPerPodContainer int) int {
+	totalEvictableContainers := 0
+	for _, pod := range evictablePods {
+		if maxPerPodContainer > 0 {
```
Checking the variable here is weird. if `maxPerPodContainer` is negative, what's the purpose of running this function at all...?

feiskyer replied:
`MaxPerPodContainer` could be `<=0`, but you are right, this check should be done outside.

yujuhong replied:
The function is only used in this file, so either the caller should perform the check or the function should do it at the beginning. I think you are checking it in the loop because you want to calculate the `totalEvictableContainers` even if `maxPerPodContainer` is a negative value. I feel like this coupling may be too much. Worst case you'll calculate all the evictable containers in a separate step. That should not be too bad.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+// * gets evictable pods with belonging containers and sandboxes which are not active and created more than gcPolicy.MinAge ago.
+// * removes oldest dead containers for each pod by enforcing gcPolicy.MaxPerPodContainer.
+// * removes oldest dead containers by enforcing gcPolicy.MaxContainers.
+// * removes dead sandboxes with zero containers.
```
Please make it clear that we don't count sandboxes when enforcing the GC policy.

feiskyer replied:
will do.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+		return err
+	}
+
+	glog.V(4).Infof("GarbageCollect gets pods %q for evicting", evictablePods)
```
nit: not sure we want to print a map as %q. Change it to %v? In reality, UIDs are enough and we don't need the rest of the map logged at V(4).

feiskyer replied:
will do.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+			containerList := evictablePods[container.podUID].containers
+			// remove the container from runtime service and also update podGCInfo.containers
+			containerIndex := evictablePods[container.podUID].removeContainer(container.containerID)
+			evictablePods[container.podUID].containers = append(containerList[:containerIndex], containerList[containerIndex+1:]...)
```
There is a lot of extra work due to the "pod" unit you use. However, so far,it seems like the "pod" unit is not really useful until `removeEvictableSandboxes`. Should we construct the "pods" only before running that? Just a thought...

yujuhong replied:
The three steps required three different eviction units:
 1. MaxPerPodContainer: `(pod UID, container name)`. Grouping by pods seems ok.
 2. MaxContainers: `(container name)`. Grouping by pods doesn't make sense.
 3. sandbox cleanup: `(sandbox id)`. Grouping by pods seems ok. 

If we can't change the order, there are some necessary transitions in between.
I suggest that we do these in the order of (1), (3), (2):
  * Both (1) and (3) can group by pods. At the end, we can convert them into a list of containers and do (2). No more funky business of deleting containers in the evictable pods.
  * Before trying to delete oldest containers globally, i.e., (2), we can make sure all the redundant sandboxes are cleaned up first.
  * The downside is that we won't be able to remove redundant sandboxes created by (2) until the next GC cycle, but I think that's fine for now.

WDYT? Does this make sense?

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+			container := allEvictableContainers[i]
+			containerList := evictablePods[container.podUID].containers
+			// remove the container from runtime service and also update podGCInfo.containers
+			containerIndex := evictablePods[container.podUID].removeContainer(container.containerID)
```
`containerIndex` could be -1 (not found), in which case there is no need to perform the append operation below.

yujuhong commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,236 @@
+			}
+		}
+
+		containerStatuses, err := m.getKubeletContainerStatuses(s.GetId())
```
This worries me. We should not need to call runtime for each individual container status. If `createdAt` is the only thing needed for garbage collection, should we include that in the information returned by `ListContainers`, as what we did for podsandbox? /cc @euank @yifan-gu

yujuhong replied:
One problem is the unnecessary performance overhead, and the other is the reliability issue. If *one* container cannot be inspected, GC will just error out. The latter is completely unacceptable. One way to alleviate this is to skip the pod, but I think adding `CreatedAt` may make more sense.

feiskyer replied:
Reasonable, GC shouldn't introduce potential performance overhead here.

feiskyer commented:
@yujuhong Addressing comments. Squashed to one commit since the data structure is re-designed. PTAL.

feiskyer commented:
Review status: 3 of 16 files reviewed at latest revision, 19 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 226 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPusZAK9pvf22ZNTXBO-r5-226:-KQdHNXjoq0gzja0hkxU:b-lzormx) ([raw file](https://github.com/kubernetes/kubernetes/blob/ff96f21cad007c06248da1c6c9fadb5878854ac5/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L226)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> The three steps required three different eviction units:
> 
>  1. MaxPerPodContainer: `(pod UID, container name)`. Grouping by pods seems ok.
> 
>  2. MaxContainers: `(container name)`. Grouping by pods doesn't make sense.
> 
>  3. sandbox cleanup: `(sandbox id)`. Grouping by pods seems ok. 
> 
> 
> 
> If we can't change the order, there are some necessary transitions in between.
> 
> I suggest that we do these in the order of (1), (3), (2):
> 
>   * Both (1) and (3) can group by pods. At the end, we can convert them into a list of containers and do (2). No more funky business of deleting containers in the evictable pods.
> 
>   * Before trying to delete oldest containers globally, i.e., (2), we can make sure all the redundant sandboxes are cleaned up first.
> 
>   * The downside is that we won't be able to remove redundant sandboxes created by (2) until the next GC cycle, but I think that's fine for now.
> 
> 
> 
> WDYT? Does this make sense?
</details>
Sorry, didn't notice this comment. The new commit doesn't work this way, it doesn't group the containers as pod again. Instead, it groups containers by `(podUID, containerName)`, so the GC policy is worked exactly with original Docker. After this, the sandboxes are filtered and only those with no containers are removed.

One drawback of this method is it calls `ListContainers()` twice, one for container GC and the other for sandbox GC (to calculate the number of containers belonging to the sandbox).

WDYT?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 226 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPusZAK9pvf22ZNTXBO-r5-226:-KQgHFBU74vWlpiNMY11:bge4daj) ([raw file](https://github.com/kubernetes/kubernetes/blob/ff96f21cad007c06248da1c6c9fadb5878854ac5/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L226)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> Sorry, didn't notice this comment. The new commit doesn't work this way, it doesn't group the containers as pod again. Instead, it groups containers by `(podUID, containerName)`, so the GC policy is worked exactly with original Docker. After this, the sandboxes are filtered and only those with no containers are removed.
> 
> One drawback of this method is it calls `ListContainers()` twice, one for container GC and the other for sandbox GC (to calculate the number of containers belonging to the sandbox).
> 
> WDYT?
</details>
That sounds reasonable. Will review later today. Thanks!

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
Rebased to fix conflicts.

feiskyer commented:
@yujuhong @yifan-gu Rebased. PTAL.

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+	for key := range cu {
+		num += len(cu[key])
+	}
+
```
nit: kill empty line

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+	}
+}
+
+// Internal information kept for containers being considered for GC.
```
godoc convention: `containerGCInfo is...`

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+}
+
+// Containers are considered for eviction as units of (UID, container name) pair.
+type evictUnit struct {
```
godoc convention: `evictUnit is ....`

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+type containersByEvictUnit map[evictUnit][]containerGCInfo
+
+// Returns the number of containers in this map.
+func (cu containersByEvictUnit) NumContainers() int {
```
godoc convension

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+
+type containersByEvictUnit map[evictUnit][]containerGCInfo
+
+// Returns the number of containers in this map.
```
godoc convention: `NumContainers returns...`

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+	return num
+}
+
+// Returns the number of pod in this map.
```
ditto

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+// Note that gc policy is not applied to sandboxes. Sandboxes are only removed when they are
+// not ready and containing no containers.
+//
+// GarbageCollect is consist of following steps:
```
s/is consist of/consists of

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+// Removes the oldest toRemove containers and returns the resulting slice.
+func (cgc *containerGC) removeOldestN(containers []containerGCInfo, toRemove int) []containerGCInfo {
+	// Remove from oldest to newest (last to first).
+	numToKeep := len(containers) - toRemove
```
nit: maybe rename it to `start`, something implies an index instead of numbers.

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+
+// enforceMaxContainersPerEvictUnit enforces MaxPerPodContainer for each evictUnit.
+func (cgc *containerGC) enforceMaxContainersPerEvictUnit(evictUnits containersByEvictUnit, MaxContainers int) {
+	for uid := range evictUnits {
```
This is not actually uid, it's uid + container name.
So maybe just rename it to `key`

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+		numContainers := evictUnits.NumContainers()
+		if numContainers > gcPolicy.MaxContainers {
+			flattened := make([]containerGCInfo, 0, numContainers)
+			for uid := range evictUnits {
```
s/uid/key

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc.go:
```diff
@@ -0,0 +1,282 @@
+			continue
+		}
+
+		// Prune out sandboxes still contains containers.
```
that still have

yifan-gu commented on pkg/kubelet/kuberuntime/kuberuntime_gc_test.go:
```diff
@@ -0,0 +1,132 @@
+		MaxPerPodContainer: 1,
+		MaxContainers:      3,
+	}, false))
+	assert.Equal(t, 3, len(fakeRuntime.Containers))
```
test the expected number of sandboxes.

yifan-gu commented:
Reviewed. Just some small nits, LGTM overall.

yujuhong commented:
Reviewed 3 of 3 files at r2, 5 of 5 files at r3, 2 of 3 files at r5, 9 of 9 files at r6, 6 of 10 files at r7.
Review status: all files reviewed at latest revision, 28 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/dockershim/convert.go, line 58 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KRW1ASoVIKYp4X-SaTt:-KRW1ASoVIKYp4X-SaTu:buf0a30) ([raw file](https://github.com/kubernetes/kubernetes/blob/d164d8ce62f449a88635f46113586bae430865dd/pkg/kubelet/dockershim/convert.go#L58)):*
> ```Go
> 	return &runtimeApi.Container{
> 		Id:        &c.ID,
> 		CreatedAt: &c.Created,
> ```

You'll likely need to fix the unit tests after rebasing.

---

*[pkg/kubelet/kuberuntime/kuberuntime_container.go, line 247 \[r7\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KRVzO3zL-U2YLAtrvyH:-KRVzO3zL-U2YLAtrvyI:bu75u8w) ([raw file](https://github.com/kubernetes/kubernetes/blob/d164d8ce62f449a88635f46113586bae430865dd/pkg/kubelet/kuberuntime/kuberuntime_container.go#L247)):*
> ```Go
> 
> 	if !kubeStatus.FinishedAt.IsZero() || kubeStatus.ExitCode != 0 {
> 		if terminationMessagePath != "" {
> ```

nit: too many indents. can just return if `terminationMessagePath` is not set.

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 134 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KPusZAK9pvf22ZNTXBO-r4-134:-KRW-QbYHJguU5Bq5RmB:b60dgs) ([raw file](https://github.com/kubernetes/kubernetes/blob/2619bed8f164aa8fbce1bee4b70f0d8b8c3a561b/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L134)):*
<details><summary><i>Previously, feiskyer (Pengfei Ni) wrote…</i></summary>
> I mean `contain` here.
</details>
I think we can reword this to make it more clear. 
`A container is evictable if it is not running and was created more than MinAge ago.`

---

*[pkg/kubelet/kuberuntime/kuberuntime_gc.go, line 62 \[r6\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326#-KQgfe9QhXIYN93PcpEN:-KQgfe9QhXIYN93PcpEO:bnllsmc) ([raw file](https://github.com/kubernetes/kubernetes/blob/88a68fa5db5add54d0d4f383326e7dde3fa0d5da/pkg/kubelet/kuberuntime/kuberuntime_gc.go#L62)):*
> ```Go
> 	// UID of the pod.
> 	uid types.UID
> ```

remove the extra line.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31326)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
@yifan-gu @yujuhong Rebased and addressed comments. PTAL.

@yujuhong btw, some of your comments are based on old commits (maybe something wrong with reviewable). Please review the latest commit.

yujuhong commented:
lgtm. Leaving up to @yifan-gu for a final round.

yifan-gu commented:
LGTM, thanks @feiskyer !

yujuhong commented:
Need to squash the commits. Removed the lgtm label.

feiskyer commented:
@yujuhong squashed.

yujuhong commented:
@feiskyer uh..more rebase.
